test/blockstore-test.js

```diff
@@ -0,0 +1,844 @@
+
+      const len = first.size + second.size + third.size - (8 * 16);
+      assert.equal(len, 128 * 16);
+
+      for (let i = 0; i < 16; i++) {
+        const expect = blocks[i];
+        const block = await store.read(expect.hash);
+        assert.bufferEqual(block, expect.block);
+      }
+    });
+
+    it('will allocate new files with block undo coins', async () => {
+      const blocks = [];
+
+      for (let i = 0; i < 16; i++) {
+        const block = random.randomBytes(128);
+        const hash = random.randomBytes(32);
+        blocks.push({hash, block});
+        await store.writeUndo(hash, block);
+        const block2 = await store.readUndo(hash);
+        assert.bufferEqual(block2, block);
+      }
+
+      const first = await fs.stat(store.filepath(types.UNDO, 0));
+      const second = await fs.stat(store.filepath(types.UNDO, 1));
+      const third = await fs.stat(store.filepath(types.UNDO, 2));
+      assert.equal(first.size, 952);
+      assert.equal(second.size, 952);
+      assert.equal(third.size, 272);
+
+      const len = first.size + second.size + third.size - (8 * 16);
+      assert.equal(len, 128 * 16);
+
+      for (let i = 0; i < 16; i++) {
+        const expect = blocks[i];
+        const block = await store.readUndo(expect.hash);
+        assert.bufferEqual(block, expect.block);
+      }
+    });
+
+    it('will recover from interrupt during block write', async () => {
+      {
+        const block = random.randomBytes(128);
+        const hash = random.randomBytes(32);
+        await store.write(hash, block);
+
+        const block2 = await store.read(hash);
+        assert.bufferEqual(block2, block);
+      }
+
+      // Manually insert a partially written block to the
+      // end of file as would be the case of an untimely
+      // interrupted write of a block. The file record
+      // would not be updated to include the used bytes and
+      // thus this data should be overwritten.
+      {
+        const filepath = store.filepath(types.BLOCK, 0);
+
+        const fd = await fs.open(filepath, 'a');
+
+        const bw = bio.write(8);
+        bw.writeU32(store.network.magic);
+        bw.writeU32(73);
+        const magic = bw.render();
+
+        const failblock = random.randomBytes(73);
+
+        const mwritten = await fs.write(fd, magic, 0, 8);
+        const bwritten = await fs.write(fd, failblock, 0, 73);
+
+        await fs.close(fd);
+
+        assert.equal(mwritten, 8);
+        assert.equal(bwritten, 73);
+      }
+
+      // Now check that this block has the correct position
+      // in the file and that it can be read correctly.
+      {
+        const block = random.randomBytes(128);
+        const hash = random.randomBytes(32);
+        await store.write(hash, block);
+
+        const block2 = await store.read(hash);
+        assert.bufferEqual(block2, block);
+      }
+    });
+
+    it('will not write blocks at the same position', (done) => {
+      let err = null;
```

Comment:
repeat for `undo`

Reply:
It's using the same code, so the double test isn't needed as it's already tested. There is an integration test between `writeUndo` and `_write`, which is covered.